harvester_e2e_tests/fixtures/terraform.py

```diff
@@ -0,0 +1,185 @@
+            with open(kubefile, "w") as f:
+                f.write(kubeconfig)
+
+            with open(self.workdir / "provider.tf", "w") as f:
+                f.write(TF_PROVIDER.format(
+                    tf_version=">=0.13", config_path=kubefile.resolve(),
+                    provider_source="harvester/harvester", provider_version=provider_version
+                ))
+
+            return self.execute("init")
+
+        def save_as(self, content, filename, ext=".tf"):
+            filepath = self.workdir / f"{filename}{ext}"
+            with open(filepath, "w") as f:
+                f.write(content)
+
+        def apply_resource(self, resource_type, resource_name):
+            return self.execute(f"apply -auto-approve -target {resource_type}.{resource_name}")
+
+        def destroy_resource(self, resource_type, resource_name):
+            return self.execute(f"destroy -auto-approve -target {resource_type}.{resource_name}")
+
+    path = Path(request.config.getoption('--terraform-scripts-location'))
+    harv = TerraformHarvester(tf_executor, path / datetime.now().strftime("%Hh%Mm_%m-%d"))
+    kuebconfig = api_client.generate_kubeconfig()
+    out, err, exc_code = harv.initial_provider(kuebconfig, tf_provider_version)
+    assert not err and 0 == exc_code
+
+    return harv
+
+
+@pytest.fixture(scope="session")
+def tf_resource(request, tf_provider_version):
+    from dataclasses import dataclass
+
+    @dataclass
+    class ResourceContext:
+        type: str
+        name: str
+        ctx: str
+
+    class TerraformResource:
+        @classmethod
+        def _support_version(cls, version):
+            return True
+
+        @classmethod
+        def from_version(cls, version):
+            for c in cls.__subclasses__():
+                if c._support_version(version):
+                    return c()
+            return cls()
+
+        def __init__(self):
+            self.executor = Path("./terraform_test_artifacts/json2hcl").resolve()
+
+        def convert_to_hcl(self, json_spec, raw=False):
+            rv = run(f"echo {json.dumps(json_spec)!r} | {self.executor!s}",
+                     shell=True, stdout=PIPE, stderr=PIPE)
+            if raw:
+                return rv
+            if rv.stderr:
+                raise TypeError(rv.stderr, rv.stdout, rv.returncode)
+            out = rv.stdout.decode()
+            out = re.sub(r'"resource"', "resource", out)  # resource should not quote
+            out = re.sub(r"\"(.+?)\" =", r"\1 =", out)  # property should not quote
+            return out
+
+        def ssh_key(self, resource_name, name, public_key, *, convert=True, **kwargs):
+            spec = {"name": name, "public_key": public_key}
+            spec.update(kwargs)
+            rv = dict(resource=dict(harvester_ssh_key={resource_name: spec}))
+
+            if convert:
+                return ResourceContext("harvester_ssh_key", resource_name, self.convert_to_hcl(rv))
+            return rv
```

Comment:
May extract `resource_type`.
```
resource_type = "harvester_ssh_key"
...
rv = {"resource":{resource_type:{resource_name: spec}}}

if convert:
    return ResourceContext(resource_type, resource_name, self.convert_to_hcl(rv))
return rv
```
Same for other resources.

Reply:
https://github.com/harvester/tests/pull/979